### ISSUE TITLE
RPC end points scaling using EFS

### ISSUE THREAD

**Issue by LeoHChen** [harmony-one/harmony#3902]
## Summary

Use EFS as the storage backend of the blockchain, and auto-scale nodes as frontends to serve RPC requests.

## Current Design

We have multiple nodes behind the RPC endpoints. Each of them running a non-archival explorer node with a full (pruned) blockchain DB to serve RPC requests.

## Problems

When we have spikes of RPC requests, the node could be overloaded by the requests and may fall out of sync with the blockchain. If it fall out of sync of the blockchain, the load balancer target group may automatically takes the node offline, which may worse the situation of other nodes in the target group.

## Proposal

The proposal is to set a long term scaling solution using auto-scaling group to respond to the increase of RPC requests.
For auto-scaling to work, the nodes have to be stateless and read the DB only. So, we propose to use one node to write to blockchain db and other nodes to do read only access to the db. The first step is to try to use EFS (elastic file system)  as the blockchain db backend, and auto-scaling group nodes to do read-only access of the EFS.

Some more research and tests are still needed for this proposal.  For example,

- [ ] One node as the writer to the EFS, and many nodes as readers
- [ ] Code modification maybe needed for read-only access to blockchain db
- [ ] Auto-scaling setup with AMI support

The ETA for this project is about 1 month.

**Comment by rlan35** [comment]
> @rlan35 @JackyWYX Can I have your input here on harmony codebase change to allow read/read redirection to a given storage disk (network / local) based on the RPC method called ?

Harmony uses leveldb for chain storage. Leveldb read and write local files. More research is needed on leveldb codebase to see whether it’s possible to read from remote db.

**Comment by sophoah** [comment]
@rlan35 maybe let me rephrase the request. Is it possible for a node to work as read-only (ie accessing the DB stored on an EFS or any filesystem) but is not writing/synching any new blocks ? That node would basically serving all the read-only RPC/API request.

@LeoHChen we might want to test both prune and archival DB, since performance will vary here due to the different RPC requests
